man/man6/wrobots.6
.TH WROBOTS 6 "Version 1, Release 4" "W Window System" "W PROGRAMS"
.SH NAME
wrobots \- Robots game for W
.SH SYNOPSIS
.nf
wrobots [-w <width>] [-h <height>]
.fi
.SH DESCRIPTION
.I WRobots
is a W version of game Daleks on Atari or Robots on unix.  In the game
you need to avoid the robots that 'zoom' onto you and to try to get them
to collision into each other (they are pretty stupid :-) or into an
already collapsed heap of robots instead.  You'll get onto next level
when you have collided all of the robots.  Game ends when one of the
robots caughts you.
.SH ARGUMENTS
.TP
.B -w
Lets you specify the play area width in blocks.
.TP
.B -h
Lets you specify the play area height in blocks.
.SH CONTROLS
You can guide your character either by clicking on a direction you want
him to move or by using the keypad numbers (1-9).  Key 't' teleports and
key 'l' can be used for 'last stand' (when you think that remaining
robots will not get you).  Note that teleporting is chancy, you can land
next to a robot...
.SH TODO
Some graphics would be nice. Currently the game looks fairly boring.
.SH LAST CHANGED
8/1996 by Eero Tamminen


wt/CHANGES
++oddie 9/2000:
- fixed listbox/filesel resizing & added minimum size setopt.

++bsittler 7/2000:
- Fixed vt color/reverse and attribute resetting
- Add HTTP/1.1 'Host' header to Wetscape requests
- Other minor changes (netgames, wvt, waudio)

++jcarlson 7/2000:
- WTINYSCREEN changes

++eero, 5/1999:
- Fixed key handling for special keys in Wt/getstring.c & wedit/wedit.c.

++eero, 1/1999:
- Added support for new W special keys so that now on W platforms which
implement special key mappings, you don't need to use Emacs shortcuts
with text widgets (or remap your console cursor keys)...


------ notes for version released with W R1PL0 ----------

ADDITIONAL WIDGETS

- Arrow, a simple button with an arrow drawn into it.

- Dial, a dial widget where you can position a needle into desired
  angle.  This isn't in the W toolkit library itself as then you would
  need to link math library to all of your Wt programs.  You'll find it
  in the warzone game tar-ball.

- Edittext, mainly intended for simple one page stuff, but as this
  spaghetti can do some additional stuff like paragraph formatting,
  there's a real abuse built on this, the WEdit.  As edittext widget is
  large and buggy, it's in the WEdit directory instead of the toolkit.

- Menu, a strip of buttons from which popups pop-up.  Both the popups
  and menus use a structure defined in Wt.h instead of strings that Kay
  was using.  The reason is that I want callbacks to follow
  automatically hierarchically opened popups.

- Popup, hierarchical popups.  Can include subpopups, images, lines and
  in addition items can also be disabled and checked.

- Range, a scrollbar composite with which you can select a value between
  given limits.

- Select, a getstring widget coupled with an arrow openable listbox from
  which items can be selected into a getstring widget.


OTHER CHANGES

- vt can now do everything wterm  can do + user can rebind the colors
  in wtrc.

- Added '.wtconfig' file and made makefiles use the W compilation
  configuration file '.config' in parent directory.

- wt_do_event() doesn't call wt_dispatch_event() anymore with
  EVENT_GADGET/GADGET_EXIT event as those might not have valid window
  pointer.  If you want to check for abnormal (unboud) exit, check the
  wt_return value.  GADGET_EXIT returns a negative number and
  GADGET_CLOSE (in shell widget) a positive number now.

- Added xterm mouse events, reverse mode, fd/bgcolor setting + improved
  double/triple click timing in VT widget.

- Aliases a couple new tags into others in HTML widget and added
  HTML header (ignored) and TITLE parsing.

- Widget event processing functions return now unrecognized or
  non-processed WEVENTs or NULL instead of long.  Assumption is that if
  function returns WEVENT, the widget whose action it was, is still
  around.  If in doubt, use NULL.  Unrecognized key events should be
  returned for focus change though.  (actually it could still have been
  an int, but returning a WEVENT offers a possibility for event mapping
  widgets.)

- Added wt_global structure containing Wt configuration variables and
  made widgets to use it.  wt_init() will now read these variables from
  ~/.wtrc + a list of (variable, value) pairs.  Widget specific
  initialization code in their init() functions may read these with the
  wt_variable() function.

- HTML widget fonts can be configured by above method.

- Added wt_loadfont() function which will use the user specified Wt
  default if widget doesn't specify some font parameter.  It also exits
  program if Wt default fonts can't be loaded, because otherwise
  unavailable fonts may chrash naive programs.

- Changed widgets to reflect the new w_loadfont() syntax (NULL/zero
  values tell Wlib/server to use server defaults).

- Added new WT_FONTSIZE define.  The earlier WT_FONT setopt sets now the
  font *family* name, new one sets the font size.

- Getopting WT_FONT returns WFONT* pointer also for above reason.

- Removed 'fontname' members from widget structures.  Needed information
  and more can be gotten from the 'font' member if needed.

- Fileselector has now separate lists for directory and file names.
  Both of course have an autolocator (type & return in addition to point
  & click).

- Wt_keys.h defines values for some special keys like cursors etc.  Half
  of the defines are Emacs style and half what the special keys itself
  give or are in use on other platforms.

- Changed Wt option defines.  There were a couple overlapping ones.


------- Kay's original notes: ----------------------------------------------

- added some missing close() and truncate the file to zero length before
  writing a new selection in selection.c.

- better selection handling in VT widget.

- made everything resizeable.

- applied patches by eero and fixed some more bugs myself.

- implemented a simple clipboard and changed getstring and vt widget to use it.

- fixed several serious bugs in the html widget.

- wrote VT widget.

- added an option to the icon editor to sync edit field with display
  field (simon).

- added missing special chars to html widget (simon).

- added code to the fileselector to set a default file name.

- lots of changes by eero (button.c, fileselect.c, listbox.c, getstring.c), some
  fixes to his changes by me.

- added missing wt_ungetfocus() to listbox and getstring.

- wrote Tk-like packer widget.

- new fileselect widget and testprogram by eero. Added better filename
  matching (eeros version was a joke, really).

- getstring and listbox updates by eero.

- changed the scrollbar widget's user interface to become more user friendly.
  See scrollbar.doc for details on how to use it. Also changed viewport and
  listbox to use the new interface.

- added listbox_focus(). Otherwise listbox cannot receive keyboard focus...

- big bugfix: when wt_delete() is called on a widget all pending timers must
  be canceled! (affected list, iconedit and getstring).


games/wrobots.c
/*
 * wrobots.c, a part of the W Window System
 *
 * Copyright (C) 1996 by Eero Tamminen
 *
 * This package is free software; you can redistribute it and/or modify it
 * under the terms specified in the docs/COPYRIGHTS file coming with this
 * package.
 *
 * - robots / daleks game for W
 *
 * TODO
 * - draw some nice bitmaps for player, robots and robot 'heap'
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <Wlib.h>

#define GAME_NAME	"W-Robots"
#define DEF_ROBOTS	8		/* robots at start */
#define MAX_ROBOTS	128		/* maximum amount of robots */
#define MIN_WIDTH	10		/* minimum play area size */
#define MIN_HEIGHT	10
#ifdef WTINYSCREEN
  #define DEF_WIDTH	19		/* play area size */
  #define DEF_HEIGHT	17
#else
  #define DEF_WIDTH	20		/* play area size */
  #define DEF_HEIGHT	20
#endif
#define SCORE		10		/* points / robots */

/* moves */
#define INVALID		0
#define MOVE		1
#define STAY_HERE	2
#define TELEPORT	3
#define LAST_STAND	4

typedef struct
{
  int x, y;
  int alive;
} ROBOTS;

static int Width = DEF_WIDTH, Height = DEF_HEIGHT;

/* game prototypes */
int  parse_args(int argc, char *argv[]);
void random_pos(int robots, ROBOTS *pos, int *x, int *y);
void init_level(int robots, ROBOTS *pos, int *manx, int *many);
int  move(int robots, ROBOTS *pos, int *manx, int *many);
int  hunt(int robots, ROBOTS *pos, int manx, int many);

/* window / event / graphics handling prototypes */
const char *win_init(void);
void clear_win(void);
void wait_input(long msecs);
int  get_dir(int *x, int *y);
int  process_input(WEVENT *ev, int *x, int *y);
void draw_rip(int x, int y, int score);
void show_info(int level, int score);
void draw_dalek(int x, int y);
void draw_heap(int x, int y);
void draw_man(int x, int y);
void clear(int x, int y);


int main(int argc, char *argv[])
{
  int level = 0, score = 0, enemies, manx, many, dir, alive = 1,
       robots, dead, dbase = DEF_ROBOTS;
  static ROBOTS pos[MAX_ROBOTS];

  if(!parse_args(argc, argv))
  {
    fprintf(stderr, "\n" GAME_NAME " w/ 1996 Eero Tamminen\n\n");
    fprintf(stderr, "Usage: %s [-w <width (blocks)>] [-h <height (blocks)>]\n", *argv);
    fprintf(stderr, "\nTo move, use either keypad or mouse.\n");
    return -1;
  }
  /* base number of robots on the play area */
  dbase = Width * Height / 48;
  srand(time(0));
  win_init();
  do
  {
    enemies = robots = dbase + level * dbase / 4;
    init_level(robots, pos, &manx, &many);
    level++;
    dir = INVALID;
    show_info(level, score);
    do
    {
      if(dir != LAST_STAND)
        dir = move(robots, pos, &manx, &many);

      dead = hunt(robots, pos, manx, many);
      score += dead * SCORE * level;
      if(dead)
        show_info(level, score);
      if(dead < 0)
	alive = 0;
      else
        enemies -= dead;
      wait_input(128);

    } while(alive && enemies);		/* player and robots alive? */

  } while(alive);

  draw_rip(manx, many, score);
  wait_input(-1);
  return 0;
}

/* get play area size */
int parse_args(int argc, char *argv[])
{
  int idx = 0;
  while(++idx < argc)
  {
    /* an one letter option and argument for it? */
    if(argv[idx][0] == '-' && argv[idx][2] == '\0' && idx+1 < argc)
      switch(argv[idx++][1])
      {
	case 'w':				/* window width */
	  if((Width = atoi(argv[idx])) < MIN_WIDTH)
	    Width = MIN_WIDTH;
	  break;
	case 'h':				/* window height */
	  if((Height = atoi(argv[idx])) < MIN_WIDTH)
	    Height = MIN_HEIGHT;
	  break;
	default:
	  return 0;
      }
    else
      return 1;
  }
  return 1;	/* success */
}

void random_pos(int robots, ROBOTS *pos, int *x, int *y)
{
  int idx;

  for(;;)
  {
    *x = rand() % Width;
    *y = rand() % Height;

    /* not onto a robots? */
    for(idx = 0; idx < robots; idx++)
      if(*x == pos[idx].x && *y == pos[idx].y)
	continue;

    return;
  }
}

/* randomize robots and player positions */
void init_level(int robots, ROBOTS *pos, int *manx, int *many)
{
  int idx;

  clear_win();
  for(idx = 0; idx < robots; idx++)
  {
    pos[idx].alive = 1;
    random_pos(idx-1, pos, &pos[idx].x, &pos[idx].y);
    draw_dalek(pos[idx].x, pos[idx].y);
  }

  random_pos(robots, pos, manx, many);
  draw_man(*manx, *many);
}

/* let the user make a move (show valid ones) */
int move(int robots, ROBOTS *pos, int *manx, int *many)
{
  int x, y, dir, idx;

  for(;;)
  {
    x = *manx;
    y = *many;
    dir = get_dir(&x, &y);
    if(dir != MOVE)
      break;

    if(x >= 0 && y >= 0 && x < Width && y < Height)
    {
      for(idx = 0; idx < robots; idx++)
      {
        if(x == pos[idx].x && y == pos[idx].y)
	  continue;
      }
      break;
    }
  }

  /* keep in place / last stand */
  if(dir == STAY_HERE || dir == LAST_STAND)
    return dir;

  clear(*manx, *many);
  if(dir == TELEPORT)
  {
    random_pos(robots, pos, manx, many);
  }
  else
  {
    *manx = x;
    *many = y;
  }

  draw_man(*manx, *many);
  return dir;
}

/* move the robots into same direction that the player is
 * return the number of collided robots or -1 if they got you
 */
int  hunt(int robots, ROBOTS *pos, int manx, int many)
{
  int test, idx, hits = 0, dead = 0;

  /* move the alive robots ones closer to the player */
  for(idx = 0; idx < robots; idx++)
    if(pos[idx].alive)
    {
      /* clear old robots from the screen */
      clear(pos[idx].x, pos[idx].y);

      /* get closer */
      if(pos[idx].x < manx)
        pos[idx].x++;
      else if(pos[idx].x > manx)
        pos[idx].x--;

      if(pos[idx].y < many)
        pos[idx].y++;
      else if(pos[idx].y > many)
        pos[idx].y--;

      /* check whether they got you... */
      if(pos[idx].x == manx && pos[idx].y == many)
	  hits--;
    }

  /* check for collisions */
  for(idx = 0; idx < robots; idx++)
    if(pos[idx].alive)
    {
      for(test = 0; test < robots; test++)
      {
        if(test != idx && pos[idx].x == pos[test].x && pos[idx].y == pos[test].y)
        {
	  if(pos[idx].alive)
	  {
	    pos[idx].alive = 0;
	    draw_heap(pos[idx].x, pos[idx].y);
	    dead++;
	  }
	  if(pos[test].alive)
	  {
	    pos[test].alive = 0;
	    dead++;
	  }
	  /* check next alive one */
	  continue;
	}
      }
      /* no collision? */
      if(pos[idx].alive)
        draw_dalek(pos[idx].x, pos[idx].y);
    }

  if(hits)
    return hits;
  return dead;
}


/* window / graphics handling */

static WWIN *PlayWin;

#define PROPERTIES	(W_MOVE | W_TITLE | W_CLOSE | EV_KEYS | EV_MOUSE)
#ifdef WTINYSCREEN
#define BLK_SIZE	8	/* one area unit size */
#else
#define BLK_SIZE	16	/* one area unit size */
#endif

/* M_DRAW is the default graphics mode */

const char *win_init(void)
{
  if(!(w_init()))
    return "can't connect W server";

  if(!(PlayWin = w_create(Width * BLK_SIZE, Height * BLK_SIZE, PROPERTIES)))
    return "unable to create game window";

  w_settitle(PlayWin, GAME_NAME);

  if(w_open(PlayWin, UNDEF, UNDEF) < 0)
    return "unable to open input window";

  return 0;
}

void clear_win(void)
{
  w_setmode(PlayWin, M_CLEAR);
  w_pbox(PlayWin, 0, 0, Width * BLK_SIZE, Height * BLK_SIZE);
  w_setmode(PlayWin, M_DRAW);
}

void wait_input(long msecs)
{
  WEVENT *ev;
  if(!(ev = w_queryevent(NULL, NULL, NULL, msecs)))
    return;

  if(ev->type == EVENT_GADGET &&
     (ev->key == GADGET_EXIT || ev->key == GADGET_CLOSE))
    exit(0);
}

/* return direction (0-8) or teleport (9) index */
int get_dir(int *x, int *y)
{
  int dir;
   while(!(dir = process_input(w_queryevent(NULL, NULL, NULL, -1), x, y)));
  return dir;
}

int process_input(WEVENT *ev, int *x, int *y)
{
  if (!ev) {
    return INVALID;
  } 
  switch(ev->type)
  {
    /* return direction or teleport index */
    case EVENT_KEY:
      switch(ev->key & 0xFF)
      {
	/* directions */
	case '8':
	  (*y)--;
	  break;
	case '2':
	  (*y)++;
	  break;
	case '4':
	  (*x)--;
	  break;
	case '6':
	  (*x)++;
	  break;
	case '7':
	  (*x)--;
	  (*y)--;
	  break;
	case '9':
	  (*x)++;
	  (*y)--;
	  break;
	case '1':
	  (*x)--;
	  (*y)++;
	  break;
	case '3':
	  (*x)++;
	  (*y)++;
	  break;

	/* other 'moves' */
	case '5':
	  return STAY_HERE;
	case 't':
	case 'T':
	  return TELEPORT;
	case 'l':
	case 'L':
	  return LAST_STAND;
	default:
	  return INVALID;
      }
      break;

    case EVENT_MPRESS:
      ev->x /= BLK_SIZE;
      ev->y /= BLK_SIZE;
      if(ev->x == *x && ev->y == *y)
        return STAY_HERE;

      if(ev->x != *x)
        ev->x < *x ? (*x)-- : (*x)++;
      if(ev->y != *y)
        ev->y < *y ? (*y)-- : (*y)++;
      break;

    case EVENT_GADGET:
     if(ev->key == GADGET_EXIT || ev->key == GADGET_CLOSE)
       exit(0);

    default:
     return INVALID;
  }
  return MOVE;
}

void show_info(int level, int score)
{
  char msg[64];
  sprintf(msg, "Level: %d Score: %d", level, score);
  w_settitle(PlayWin, msg);
}


/* these could of course use bitmaps... I decided to try drawing
 * something abstract for a change <g>	- Eero
 */

void draw_dalek(int x, int y)
{
  w_pbox(PlayWin, x*BLK_SIZE + 2, y*BLK_SIZE + 2, BLK_SIZE - 3, BLK_SIZE - 3);
}

void draw_heap(int x, int y)
{
  w_dpbox(PlayWin, x*BLK_SIZE + 1, y*BLK_SIZE + 1, BLK_SIZE - 2, BLK_SIZE - 2);
}

void draw_man(int x, int y)
{
  w_setmode(PlayWin, M_INVERS);
  w_pbox(PlayWin, x*BLK_SIZE + 3, y*BLK_SIZE + 3, BLK_SIZE - 5, BLK_SIZE - 5);
  w_pbox(PlayWin, x*BLK_SIZE + 5, y*BLK_SIZE + 5, BLK_SIZE - 9, BLK_SIZE - 9);
  w_setmode(PlayWin, M_DRAW);
}

void draw_rip(int x, int y, int score)
{
  char msg[64];

  w_setmode(PlayWin, M_CLEAR);
  w_pbox(PlayWin, x*BLK_SIZE, y*BLK_SIZE, BLK_SIZE, BLK_SIZE);
  w_setmode(PlayWin, M_INVERS);
  w_dpbox(PlayWin, x*BLK_SIZE + 3, y*BLK_SIZE + 3, BLK_SIZE - 5, BLK_SIZE - 5);
  w_dpbox(PlayWin, x*BLK_SIZE + 5, y*BLK_SIZE + 5, BLK_SIZE - 9, BLK_SIZE - 9);
  sprintf(msg, "You die... Score: %d", score);
  w_settitle(PlayWin, msg);
}

void clear(int x, int y)
{
  w_setmode(PlayWin, M_CLEAR);
  w_pbox(PlayWin, x*BLK_SIZE, y*BLK_SIZE, BLK_SIZE, BLK_SIZE);
  w_setmode(PlayWin, M_DRAW);
}
